File: fs-api/fs_api.go
package fs_api

//#cgo CFLAGS: -I/usr/include/glusterfs -I./
//#cgo LDFLAGS: -L/usr/local/lib/ -lgfapi
/*
#define _GNU_SOURCE
#include "fs_api.h"
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <glusterfs/api/glfs.h>
#include <glusterfs/api/glfs-handles.h>
*/
import (
	"C"
)
import (
	"errors"
	"fmt"
	"os"
	"unsafe"
)

type FsApi struct {
	api     *C.fs_api
	address string
}
type FsFd struct {
	Fd *C.fs_fd
}

//create glusterfs handler base on fstype
func NewFsApi(volume, addr string, port int, fstype bool) (*FsApi, error) {
	if volume == "" || addr == "" || port <= 0 {
		return nil, fmt.Errorf("the parameter is null")
	}
	var fsApi *FsApi
	var err error
	if fstype {
		cvolume := C.CString(volume)
		caddr := C.CString(addr)
		defer C.free(unsafe.Pointer(cvolume))
		defer C.free(unsafe.Pointer(caddr))
		api := C.fs_api_init(cvolume, caddr, C.int(port))
		if api == nil {
			err = errors.New("init glfs failed")
		}
		fsApi = &FsApi{
			address: fmt.Sprintf("%s:%d:%s", addr, port, volume),
			api:     api,
		}
	}
	return fsApi, err
}

//check file state
func (fsApi *FsApi) Stat(name string) error {
	var st C.struct_stat
	cname := C.CString(name)
	defer C.free(unsafe.Pointer(cname))
	ret, err := C.fs_api_stat(fsApi.api, cname, &st)
	if int(ret) != 0 || err != nil {
		return err
	}
	return nil
}
func (fsApi *FsApi) RmFile(path string) error {
	cpath := C.CString(path)
	defer C.free(unsafe.Pointer(cpath))
	ret, err := C.fs_api_rmfile(fsApi.api, cpath)
	if int(ret) != 0 || err != nil {
		return err
	}
	return nil
}
func (fsApi *FsApi) RmDir(path string) error {
	cpath := C.CString(path)
	defer C.free(unsafe.Pointer(cpath))
	ret, err := C.fs_api_rmdir(fsApi.api, cpath)
	if int(ret) != 0 || err != nil {
		return err
	}
	return nil
}
func (fsApi *FsApi) Open(filename string, flags int) (*FsFd, error) {
	var err error
	fd := &FsFd{}
	cfilename := C.CString(filename)
	defer C.free(unsafe.Pointer(cfilename))
	if _, err := C.fs_api_open(fsApi.api, &fd.Fd, cfilename, C.int(flags)); err != nil {
		return nil, err
	}
	return fd, err
}
func (fsApi *FsApi) Creat(filename string, flags int, mode os.FileMode) (*FsFd, error) {
	fd := &FsFd{}
	cfilename := C.CString(filename)
	defer C.free(unsafe.Pointer(cfilename))
	ret, err := C.fs_api_creat(fsApi.api, &fd.Fd, cfilename, C.int(flags), C.mode_t(mode))
	if int(ret) != 0 {
		return nil, err
	}
	return fd, nil
}

func (fsApi *FsApi) Mkdir(path string, mode os.FileMode) error {
	cpath := C.CString(path)
	defer C.free(unsafe.Pointer(cpath))
	ret, err := C.fs_api_mkdir(fsApi.api, cpath, C.mode_t(mode))
	if int(ret) != 0 {
		return err
	}
	return nil
}
func (fsApi *FsApi) RmAllFileFromPath(path string) error {
	cpath := C.CString(path)
	defer C.free(unsafe.Pointer(cpath))
	ret, err := C.fs_api_rm_file_from_path(fsApi.api, cpath)
	if int(ret) != 0 {
		return err
	}
	return nil
}
func (fsApi *FsApi) Seek(fd *FsFd, offset int64, whence int) (*FsFd, error) {
	coffset := C.off_t(offset)
	if fd == nil {
		return nil, errors.New("fd is nil")
	}
	ret, err := C.fs_api_seek(fsApi.api, fd.Fd, coffset, C.int(whence))
	if int(ret) != 0 {
		return nil, err
	}
	return fd, nil
}
func (fsApi *FsApi) Write(fd *FsFd, data []byte) (int64, error) {
	sz, err := C.fs_api_write(fsApi.api, fd.Fd, unsafe.Pointer(&data[0]), (C.size_t)(len(data)))
	if err != nil {
		return int64(-1), err
	}
	return int64(sz), nil
}

func (fsApi *FsApi) Read(fd *FsFd, data []byte) (int64, error) {
	sz, err := C.fs_api_read(fsApi.api, fd.Fd, unsafe.Pointer(&data[0]), (C.size_t)(len(data)))
	if err != nil {
		return int64(-1), err
	}
	return int64(sz), nil
}
func (fsApi *FsApi) Close(fd *FsFd) error {
	var err error
	if fd != nil {
		_, err = C.fs_api_close(fd.Fd)
	}
	return err
}
func (fd *FsFd)Size() (int64,error) {
	return 0,nil
}

func (fsApi *FsApi) Releae() {
	if fsApi != nil {
		C.fs_api_deinit(fsApi.api)
	}
}

File: fs-api/fs_api.c
/*************************************************************************
    > File Name: fs_api.c
  > Author:perrynzhou
  > Mail:perrynzhou@gmail.com
  > Created Time: Wednesday, September 16, 2020 PM03:11:03
 ************************************************************************/

#include "fs_api.h"
#include <fcntl.h>
#include <glusterfs/api/glfs-handles.h>
#include <glusterfs/api/glfs.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <stdio.h>
fs_api *fs_api_init(char *volume, char *addr, int port)
{
  fs_api *fapi = calloc(1, sizeof(fs_api *));
  remove("/tmp/fusion-storage-gateway.log");
  char buf[1024] = {'\0'};
  snprintf((char *)&buf, 128, "/tmp/fusion-storage-gateway.log");
  glfs_t *fs = glfs_new(volume);
  glfs_set_volfile_server(fs, "tcp", addr, port);
  glfs_set_logging(fs, (char *)&buf, 9);
  glfs_init(fs);
  fapi->fs = fs;
  return fapi;
}
off_t fs_api_seek(fs_api *fapi, fs_fd *fd, off_t offset, int whence)
{
  if (fd->lfd == -1 || fd->gfd == NULL)
  {
    return -1;
  }
  if (fapi == NULL)
  {
    return lseek(fd->lfd, offset, whence);
  }
  return glfs_lseek(fd->gfd, offset, whence);
}
int fs_api_stat(fs_api *fapi, const char *pathname, struct stat *st)
{
  if (pathname == NULL)
  {
    return -1;
  }
  if (fapi == NULL)
  {
    return stat(pathname, st);
  }
  return glfs_stat(fapi->fs, pathname, st);
}
int fs_api_open(fs_api *fapi, fs_fd **fd_ptr, const char *pathname, int flags)
{
  if (*fd_ptr == NULL)
  {
    *fd_ptr = calloc(1, sizeof(fs_fd *));
  }
  fs_fd *fd = *fd_ptr;
  if (fd == NULL)
  {
    return -1;
  }
  fd->lfd = -1;
  fd->gfd = NULL;
  if (fapi == NULL)
  {
    fd->lfd = open(pathname, flags);
    return fd->lfd;
  }
  //not support create file
  //flags only O_RDWR
  fd->gfd = glfs_open(fapi->fs, pathname, flags);
  if (fd->gfd == NULL)
  {
    return -1;
  }
  return 0;
}
int fs_api_rm_file_from_path(fs_api *fapi, const char *path)
{
  char buffer[4096] = {'\0'};
  char dirent_buffer[512] = {'\0'};
  struct dirent *dt = NULL;
  glfs_fd_t *gfd = glfs_opendir(fapi->fs, path);
  while (glfs_readdir_r(gfd, (struct dirent *)dirent_buffer, &dt), dt)
  {
    size_t len = strlen(dt->d_name);
    snprintf((char *)&buffer, 4096, "/%s/%s", path, dt->d_name);
    if (dt->d_type == DT_REG)
    {
      glfs_unlink(fapi->fs, (char *)&buffer);
    }
  }
  return 0;
}
ssize_t fs_api_read(fs_api *fapi, fs_fd *fd, void *buf, size_t count)
{
  if (fapi == NULL && fd->lfd != -1)
  {
    return read(fd->lfd, buf, count);
  }
  return glfs_read(fd->gfd, buf, count, 0);
}
ssize_t fs_api_write(fs_api *fapi, fs_fd *fd, void *buf, size_t count)
{
  if (fapi == NULL && fd->lfd != -1)
  {
    return write(fd->lfd, buf, count);
  }
  return glfs_write(fd->gfd, buf, count, 0);
}
void fs_api_close(fs_fd *fd)
{
  if (fd != NULL && fd->gfd != NULL)
  {
    glfs_close(fd->gfd);

    if (fd != NULL)
    {
      free(fd);
    }
    if (fd != NULL)
    {
      free(fd);
    }
    return;
  }
}
int fs_api_creat(fs_api *fapi, fs_fd **fd_ptr, const char *pathname, int flags, mode_t mode)
{
  if (*fd_ptr == NULL)
  {
    *fd_ptr = calloc(1, sizeof(fs_fd *));
  }
  fs_fd *fd = *fd_ptr;
  if (fd == NULL)
  {
    return -1;
  }
  if (fapi == NULL)
  {
    fd->lfd = creat(pathname, mode);
    return fd->lfd;
  }
  if ((fd->gfd = glfs_creat(fapi->fs, pathname, flags, mode)) == NULL)
  {
    return -1;
  }
  return 0;
}

int fs_api_mkdir(fs_api *fapi, const char *path, mode_t mode)
{
  if (fapi != NULL)
  {
    return glfs_mkdir(fapi->fs, path, mode);
  }
  return mkdir(path, mode);
}
int fs_api_rmfile(fs_api *fapi, const char *path)
{
  if (fapi == NULL)
  {
    return remove(path);
  }
  return glfs_unlink(fapi->fs, path);
}
int fs_api_rmdir(fs_api *fapi, const char *path)
{
  if (fapi == NULL)
  {
    return rmdir(path);
  }
  return glfs_rmdir(fapi->fs, path);
}
void fs_api_deinit(fs_api *fapi)
{
  if (fapi != NULL)
  {
    glfs_fini(fapi->fs);
    free(fapi);
    fapi = NULL;
  }
}

#ifdef FS_API_TEST
int main(int argc, char *argv[])
{
  //172.25.78.19:/train_vol
  //172.25.78.11:rep_ssd_vol
  //"test_volume", "10.193.51.144"
  if (argc < 3)
  {
    fprintf(stdout, "usage:%s {volume} {host}\n", argv[0]);
    exit(-1);
  }
  fs_api *fapi = fs_api_init(argv[1], argv[2], 24007);
  if (fapi == NULL)
  {
    fprintf(stdout, "inint failed\n");
    return -1;
  }

  char buf[64] = "1";
  char *test_file = strdup((char *)&buf);
  fs_fd *fd = NULL;
  fprintf(stdout, "fs_api_open:%d\n", fs_api_open(fapi, &fd, test_file, O_RDWR));
  fprintf(stdout, "write ret:%ld\n", fs_api_write(fapi, fd, (char *)&buf, blen));
  char rb[4096] = {'\0'};
  fprintf(stdout, "read ret:%ld\n", fs_api_read(fapi, fd, (char *)&rb, 4096));
  fprintf(stdout, "buf=%s\n", (char *)&rb);

  fs_api_close(fd);
  free(test_file);
}
#endif